python/arcticdb/diagnostics_engine/version_list_diagnosis.py

```diff
@@ -0,0 +1,342 @@
+    """
+    Get all the compacted versions after the last node from the version list.
+    """
+    compacted_versions = []
+    for version in parsed_versions.keys():
+        if parsed_versions[version]["deleted"] == False and version <= last_version_in_list:
+            compacted_versions.append(version)
+    return compacted_versions
+
+
+def determine_where_version_exists(present_in_version_list, parsed_versions, target):
+    """
+    Determine the existence of the target version in various locations: version list, compacted versions, snapshots.
+    """
+    exist_in_version_list = False
+    exists_as_compacted_version = False
+    exists_in_snaphots = False
+    is_tombstoned = False
+
+    if target in parsed_versions.keys():
+        # if it exists either in the version list or compacted, we don't about about if also exists in snapshots
+        if parsed_versions[target]["deleted"] == False:
+            if present_in_version_list:
+                exist_in_version_list = True
+            else:
+                exists_as_compacted_version = True
+        else:
+            if parsed_versions[target]["snapshots"]:
+                exists_in_snaphots = True
+    elif present_in_version_list:
+        is_tombstoned = True
+    return exist_in_version_list, exists_as_compacted_version, exists_in_snaphots, is_tombstoned
+
+
+def version_list_diagnosis(lib, symbol, as_of=None):
+    """
+    Perform a diagnosis on the version list of a given symbol in a library.
+
+    This function diagnoses where a specific version of a symbol exists within a library.
+    The version could be in the version list, compacted versions, or snapshots. It also checks if
+    the version is tombstoned (deleted). The function generates a detailed report on the location
+    and accessibility of the target version and visualizes the version chain using a graph.
+
+    Parameters
+    ----------
+    lib : object
+        The library object to perform the diagnosis on.
+    symbol : str
+        The symbol whose versions are to be diagnosed within the library.
+    as_of : int, optional
+        The specific version to diagnose. If not provided, the function will use the latest version.
+
+    Returns
+    -------
+    None
+        This function doesn't return anything.
+        It generates a report and a graph visualization which are displayed but not returned.
+    """
+    lib = check_and_adapt_library(lib)
+    if lib is None:
+        return
+    if not check_symbol_exists(lib, symbol):
+        return
+
+    lib_tool = LibraryTool(lib._library)
+
+    # get non-deleted versions
+    versions = lib.list_versions(symbol)
+    parsed_versions = parse_versions(versions)
+
+    # get actual version keys chain
+    keys = lib_tool.find_keys_for_id(KeyType.VERSION, symbol)
+    parsed_keys = parse_keys(keys)
+
+    # get specific version we are looking for
+    target = get_target_version(parsed_versions, parsed_keys, as_of)
```

Comment:
What if `as_of` is a timestamp? Or a snapshot name?

Reply:
To cut down the complexity of the task, it has been intentionally left unsupported.

Reply:
Done the corresponding changes to support these cases in a separate (the last one) commit